lib/haskell.nix

```diff
@@ -1,99 +1,85 @@
-                ├───"package:lib:package": package 'package-lib-package-1.0.0'
-                └───"package:test:test": package 'test-test-1.0.0'
-  */
-  makeFlake = haskellNix: projectF: args@{ ghcVersions, ... }:
-    let
-      args' = builtins.removeAttrs args [ "ghcVersions" ];
-
-      flakeForGhc = ghcVersion:
-        let
-          compiler-nix-name =
-            if ghcVersion == null
-            then null
-            else "ghc${ghcVersion}";
-          project = projectF (args' // { inherit compiler-nix-name; });
-          prefix =
-              if compiler-nix-name == null
-              then ""
-              else "${compiler-nix-name}:";
-          fixFlakeOutput = name: output:
-            if name == "devShell"
-            then output
-            else lib.mapAttrs' (n: v: lib.nameValuePair (prefix + n) v) output;
-          fixFlake =
-            lib.mapAttrs fixFlakeOutput;
-        in
-          fixFlake (project.flake {});
-
-      combineOutputs = name:
-        if name == "devShell"
-        then lib.last
-        else lib.foldl (l: r: l // r) {};
-
-    in lib.zipAttrsWith combineOutputs (map flakeForGhc (ghcVersions ++ [null]));
-
+  makeCI = haskellPkgs: {
+    # haskell project root
+    src,
+    # haskell package names
+    packageNames,
+    # cabal file with ghc versions specified in test-with
+    cabalFile,
+    # whether to build the project with stack, disable if you are not using stack
+    buildWithStack ? true,  
+    # stack files to use in addition to stack.yaml
+    stackFiles ? [],
+    # stack resolvers for building the project, they will be replaced in stack.yaml
+    resolvers ? [],
+    # extra haskell.nix arguments
+    extraArgs ? {}
+  }: let
+    pkgs = nixpkgs.legacyPackages.${haskellPkgs.system};
+    replaceDots = builtins.replaceStrings ["."] ["-"];
+
+    # invoke haskell.nix for every ghc specified in tested-with stanza of cabalFile
+    ghc-versions = getTestedWithVersions cabalFile;
+    pkgs-per-ghc = lib.genAttrs ghc-versions
+      (ghc: haskellPkgs.haskell-nix.cabalProject ({
+        inherit src;
+        compiler-nix-name = ghc;
+      } // extraArgs));
+
+    # invoke haskell.nix for stack.yaml and every file from stackFiles
+    stackYamls = lib.optionals buildWithStack ([ "stack.yaml" ] ++ stackFiles);
+    pkgs-per-stack-yaml = lib.mapAttrs' (n: v: lib.nameValuePair (replaceDots n) v)
+      (lib.genAttrs stackYamls
+        (stackYaml: haskellPkgs.haskell-nix.stackProject {
+          inherit src stackYaml;
+        } // extraArgs));
+
+    # invoke haskell.nix for every resolver specified in resolvers
+    stackResolvers = lib.optionals buildWithStack resolvers;
+    pkgs-per-resolver = lib.mapAttrs' (n: v: lib.nameValuePair (replaceDots n) v)
+      (lib.genAttrs stackResolvers
+        (resolver: haskellPkgs.haskell-nix.stackProject {
+          src = pkgs.runCommand "change resolver" { } ''
+            mkdir -p $out
+            cp -r ${src}/* .
```

Comment:
AFAIU, this won't copy hidden files and directories (whose names start from `.`). Perhaps, this won't cause issues in 95% use-cases, but still

Reply:
Perhaps, you can do `cp -r ${src} $out` and patch `stack.yaml` directly in `$out`